### PR TITLE
More tolerance for notifications keep alive timeout

### PR DESCRIPTION
**Pull request by tomaka**
I see this situation arising in the logs:

- The behaviour sends `Open` to the handler.
- The handler answers with `OpenResultErr`.
- The behaviour receives the `OpenResultErr` and immediately tries again by sending `Open` again.
- However, since the handler has transitioned to the `Closed` state after returning `OpenResultErr`, its keep-alive timeout kicks in and shuts down the connection.

While it is questionable to immediately try opening the same connection immediately again it has failed to open, it is clearly not intended that the keep-alive timeout triggers in the millisecond or so during which the handler is very temporarily in the `Closed` state.

This PR modifies the tolerance level in order to wait for 5 seconds after entering the `Closed` state, instead of just the first time after opening a connection.

### CODE REVIEW

**Review by mxinden — APPROVED** [review]

File: client/network/src/protocol/generic_proto/handler.rs [hunk 1]
```diff
@@ -167,6 +163,9 @@ enum State {
 		/// a boolean indicating whether an outgoing substream is still in the process of being
 		/// opened.
 		pending_opening: Vec<bool>,
+
+		/// When the `Closed` state was entered.
+		when_closed: Instant,
```
Comment (romanb) [inline client/network/src/protocol/generic_proto/handler.rs:168]:
```suggestion
		closed_at: Instant,
```

**Review by romanb — APPROVED** [review]
So this is mainly an optimisation to avoid closing and re-opening connections or does the right "timing" in this sense actually cause a problem? The latter would be unfortunate, as no choice for the timeout could then guarantee that the problem does not arise, if I understand correctly.

**Comment by tomaka** [comment]
Yes, this is intended as an optimization.

There *is* indeed a problem happening, which is why an `OpenResultErr` is received. But this problem is caused by https://github.com/paritytech/substrate/pull/7638 on the other side of the connection, and not by the connection being closed.

**Comment by tomaka** [comment]
bot merge

**Comment by tomaka** [comment]
`reconnect_after_disconnect` stalled on CI. The test passes locally for me. Restarted CI to see whether it's a heisenbug.

**Comment by tomaka** [comment]
I managed to reproduce the stalling locally after a few attempts.

```
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM => Connect(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY")): Starting to connect
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p <= Dial PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY")
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p => Connected(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY, Dialer { address: "/memory/5808620719642359495" }): Connection was requested by PSM.
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY"), ConnectionId(TaskId(0))) <= Open
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p => Connected(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL, Listener { local_addr: "/memory/5808620719642359495", send_back_addr: "/memory/11861255003643154226" }, ConnectionId(TaskId(0))): Not requested by PSM, disabling.
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(PeerId("12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL"), ConnectionId(TaskId(0))]) => OpenDesiredByRemote
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM <= Incoming(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL, IncomingIndex(0)).
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM => Accept(IncomingIndex(0), PeerId("12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL")): Enabling connections.
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(PeerId("12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL"), ConnectionId(TaskId(0))) <= Open
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY, ConnectionId(TaskId(0))) => OpenResultOk
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] External API <= Open(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY"))
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL, ConnectionId(TaskId(0))) => OpenResultOk
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] External API <= Open(PeerId("12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL"))
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] External API => Disconnect PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY")
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM <= Dropped(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY"))
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] External API <= Closed(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY)
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY"), ConnectionId(TaskId(0))) <= Close
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM => Connect(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY")): No connection in proper state. Delaying.
[2020-12-01T08:08:48Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY, ConnectionId(TaskId(0))) => CloseResult
[2020-12-01T08:08:53Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY"), ConnectionId(TaskId(0))) <= Open (ban expired)
[2020-12-01T08:08:53Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p => Disconnected(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY, ConnectionId(TaskId(0))): Enabled.
[2020-12-01T08:08:53Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM <= Dropped(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY)
[2020-12-01T08:08:53Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM => Connect(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY")): Will start to connect at until Instant { tv_sec: 1877, tv_nsec: 596328561 }
[2020-12-01T08:08:53Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p => Disconnected(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL, ConnectionId(TaskId(0))): Enabled.
[2020-12-01T08:08:53Z DEBUG sc_network::protocol::generic_proto::behaviour] External API <= Closed(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL)
[2020-12-01T08:08:53Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM <= Dropped(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL)
[2020-12-01T08:08:53Z DEBUG sc_network::protocol::generic_proto::behaviour] PSM => Connect(PeerId("12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL")): Will start to connect at until Instant { tv_sec: 1877, tv_nsec: 599123776 }
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p <= Dial PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY") now that ban has expired
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p <= Dial PeerId("12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL") now that ban has expired
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p => Connected(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL, Listener { local_addr: "/memory/5808620719642359495", send_back_addr: "/memory/2725835505436750365" }): Connection was requested by PSM.
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(PeerId("12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL"), ConnectionId(TaskId(1))) <= Open
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p => Connected(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY, Dialer { address: "/memory/5808620719642359495" }): Connection was requested by PSM.
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY"), ConnectionId(TaskId(1))) <= Open
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p => Connected(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY, Listener { local_addr: "/memory/15733688341532861893", send_back_addr: "/memory/13987782094582101496" }, ConnectionId(TaskId(2))): Secondary connection. Leaving closed.
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Libp2p => Connected(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL, Dialer { address: "/memory/15733688341532861893" }, ConnectionId(TaskId(2))): Secondary connection. Leaving closed.
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL, ConnectionId(TaskId(1))) => OpenResultOk
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] External API <= Open(PeerId("12D3KooWKc66QijnDrUNSVKdwuG51QNdi7uDjzbct7XsSfukvJrL"))
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] Handler(12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY, ConnectionId(TaskId(1))) => OpenResultOk
[2020-12-01T08:09:02Z DEBUG sc_network::protocol::generic_proto::behaviour] External API <= Open(PeerId("12D3KooWNuQdCtp98Q1YZBV6ifkLj8qoVMDi3ZW9hASr1EUEwdiY"))
```

**Comment by tomaka** [comment]
*Of course* the test passes on CI if I had logging.

**Comment by tomaka** [comment]
I just ran the test 100 times successfully in a row on my machine.

**Comment by gnunicorn** [comment]
is this still being worked on?

**Comment by tomaka** [comment]
No, I'm completely clueless as to why CI fails.
All tests pass smoothly locally. CI becomes green if I add any sort of logging.

**Comment by gnunicorn** [comment]
closing for a lack of progress.

**Pull request closed without merging**